web-common/src/features/entity-management/file-artifacts-store.ts

```diff
@@ -1,131 +1,241 @@
-import { derived, Readable, writable } from "svelte/store";
-import { parseDocument } from "yaml";
-import type { MetricsConfig } from "../metrics-views/metrics-internal-store";
-
-export type FileArtifactsData = {
-  // name used for the file. will not always be dependent on the file
-  name?: string;
-  errors?: Array<V1ReconcileError>;
-  jsonRepresentation?: MetricsConfig | Record<string, never>;
-  isReconciling?: boolean;
-};
-
-/**
- * Store to save data for each file artifact.
- * Currently, stores reconcile errors
- */
-export type FileArtifactsState = {
-  entities: Record<string, FileArtifactsData>;
-};
-const { update, subscribe } = writable({
-  entities: {},
-} as FileArtifactsState);
-
-const createOrUpdateFileArtifact = (
-  path: string,
-  callback: (entityData: FileArtifactsData) => void,
-) => {
-  update((state) => {
-    if (!state.entities[path]) {
-      state.entities[path] = {
-        errors: [],
-        jsonRepresentation: {},
-        isReconciling: false,
-      };
+import { removeLeadingSlash } from "@rilldata/web-common/features/entity-management/entity-mappers";
+import {
+  ResourceKind,
+  useProjectParser,
+  useResource,
+} from "@rilldata/web-common/features/entity-management/resource-selectors";
+import { getSubStore } from "@rilldata/web-common/lib/getSubStore";
+import {
+  runtimeServiceListResources,
+  type V1ParseError,
+  V1ReconcileStatus,
+  type V1Resource,
+  type V1ResourceName,
+} from "@rilldata/web-common/runtime-client";
+import type { QueryClient } from "@tanstack/svelte-query";
+import { derived, get, type Readable, writable } from "svelte/store";
+
+export class FileArtifact {
+  public readonly path: string;
+
+  public readonly name = writable<V1ResourceName | undefined>(undefined);
+
+  /**
+   * Last time the state of the resource `kind/name` was updated
+   * used to make sure we do not have unnecessary refreshes
+   */
+  public lastStateUpdatedOn: string | undefined;
+
+  public constructor(filePath: string) {
+    this.path = filePath;
+  }
+
+  public updateResource(resource: V1Resource) {
+    const curName = get(this.name);
+    if (
+      curName?.name !== resource.meta?.name?.name ||
+      curName?.kind !== resource.meta?.name?.kind
+    ) {
+      this.name.set({
+        kind: resource.meta?.name?.kind,
+        name: resource.meta?.name?.name,
+      });
+    }
+    this.lastStateUpdatedOn = resource.meta?.stateUpdatedOn;
+  }
+
+  public deleteResource() {
+    this.name.set(undefined);
+  }
+}
+
+export class FileArtifactsStore {
+  /**
+   * Map of all files and whether it is reconciling or not.
+   * If an entry is present here then there should be on in {@link artifacts} as well
+   */
+  public readonly files = writable<Record<string, boolean>>({});
+
+  /**
+   * Map of all files and its individual store
+   */
+  private readonly artifacts: Record<string, FileArtifact> = {};
+
+  // Actions
+
+  public async init(instanceId: string) {
```

Comment:
Good idea.

Reply:
Actually lets keep this as init. Mainly because the instance of fileArtifactStore could be shared. Also because init will be an async method so it will have to be called in the constructor anyway.

Reply:
Ah the async concern makes sense. I wonder if there's anything to learn from how we're [instantiating a feature flag listener in the feature flag store's constructor](https://github.com/rilldata/rill/blob/main/web-common/src/features/feature-flags.ts#L53-L76).

Reply:
We wont need continuous reactivity feature flags has, since the updates are already streamed in.